src/backend/distributed/utils/shard_pruning.c

```diff
@@ -0,0 +1,1239 @@
+		}
+
+		return false;
+	}
+
+	return expression_tree_walker(node, PrunableExpressionsWalker, context);
+}
+
+
+/*
+ * ShardArrayToList builds a list of out the array of ShardInterval*.
+ */
+static List *
+ShardArrayToList(ShardInterval **shardArray, int length)
+{
+	List *shardIntervalList = NIL;
+	int shardIndex;
+
+	for (shardIndex = 0; shardIndex < length; shardIndex++)
+	{
+		ShardInterval *shardInterval =
+			shardArray[shardIndex];
+		shardIntervalList = lappend(shardIntervalList, shardInterval);
+	}
+
+	return shardIntervalList;
+}
+
+
+/*
+ * PruneOne returns all shards in the table that match a single
+ * PruningInstance.
+ */
+static List *
+PruneOne(DistTableCacheEntry *cacheEntry, ClauseWalkerContext *context,
+		 PruningInstance *prune)
+{
+	ShardInterval *shardInterval = NULL;
+
+	/* Well, if life always were this easy... */
+	if (prune->alwaysFalse)
+	{
+		return NIL;
+	}
+
+	/*
+	 * For an equal constraints, if there's no overlapping shards (always the
+	 * case for hash and range partitioning, sometimes for append), can
+	 * perform binary search for the right interval.  That's usually the
+	 * fastest, so try that first.
+	 */
+	if (prune->equalConsts &&
+		!cacheEntry->hasOverlappingShardInterval)
+	{
+		shardInterval = FastShardPruning(cacheEntry, prune->equalConsts->constvalue);
+
+		/*
+		 * If pruned down to nothing, we're done. Otherwise see if other
+		 * methods prune down further / to nothing.
+		 */
+		if (!shardInterval)
+		{
+			return NIL;
+		}
+	}
+
+	/*
+	 * If the hash value we're looking for is known, we can search for the
+	 * interval directly.  That's fast and should only ever be the case for a
+	 * hash-partitioned table.
+	 */
+	if (prune->hashedEqualConsts)
+	{
+		int shardIndex = INVALID_SHARD_INDEX;
+		ShardInterval **sortedShardIntervalArray = cacheEntry->sortedShardIntervalArray;
+
+		Assert(context->partitionMethod == DISTRIBUTE_BY_HASH);
+
+		shardIndex = FindShardIntervalIndex(prune->hashedEqualConsts->constvalue,
+											sortedShardIntervalArray,
+											cacheEntry->shardIntervalArrayLength,
+											context->partitionMethod,
+											cacheEntry->shardIntervalCompareFunction,
+											cacheEntry->hasUniformHashDistribution);
+
+		if (shardIndex == INVALID_SHARD_INDEX)
+		{
+			return NIL;
+		}
+		else if (shardInterval &&
+				 sortedShardIntervalArray[shardIndex]->shardId != shardInterval->shardId)
+		{
+			return NIL;
+		}
+		else
+		{
+			return list_make1(sortedShardIntervalArray[shardIndex]);
+		}
+	}
+
+	/*
+	 * FIXME: We could perform ExhaustivePrune() on the individual
+	 * shardInterval, in case there's both an = and a < / >? It'd probably be
+	 * better to pre-compute that by checking a PruningInstance's
+	 * less[Equal]Consts/greater[Equal]Consts against equalConsts.
+	 */
+	if (shardInterval)
+	{
+		return list_make1(shardInterval);
+	}
+
+	/*
+	 * Should never get here for hashing, either we've filtered down to zero
+	 * or one shard, and returned.
```

Comment:
I agree, but why do we call `HasOverlappingShardInterval` then?

(we should just get rid of that function)